### ISSUE TITLE
error with alevin using long read transcriptome

### ISSUE THREAD

**Issue by moschmi** [COMBINE-lab/salmon#540]
I am analyzing single cell RNAseq with10x chemistry.
I have my own transcriptome from long read sequencing and followed all the steps, prepare metadata, make the gentrome.fasta, indexing, using bioawk for the transcript to gene table.
Getting this error consistently:
```
salmon alevin -l ISR -1 Novogene/fastq/Alin_neg/Alin_neg_S2_L008_R1_001.fastq.gz -2 Novogene/fastq/Alin_neg/Alin_neg_S2_L008_R2_001.fastq.gz --chromiumV3  -i salmon_alevin/Alin_neg_salmon_index -p 15 -o Alin_neg_alevin_output --tgMap PacBio/single_cell_pipeline/sqanti3_output/Alin_neg_txp2gene.tsv

Version Info: This is the most recent version of salmon.
Logs will be written to Alin_neg_alevin_output/logs
[2020-06-22 22:51:19.065] [jointLog] [info] setting maxHashResizeThreads to 15
[2020-06-22 22:51:19.065] [jointLog] [info] Fragment incompatibility prior below threshold.  Incompatible fragments will be ignored.
[2020-06-22 22:51:19.065] [jointLog] [info] The --mimicBT2, --mimicStrictBT2 and --hardFilter flags imply mapping validation (--validateMappings). Enabling mapping validation.
[2020-06-22 22:51:19.065] [jointLog] [info] Usage of --validateMappings implies use of minScoreFraction. Since not explicitly specified, it is being set to 0.65
[2020-06-22 22:51:19.065] [jointLog] [info] The use of range-factorized equivalence classes does not make sense in conjunction with --hardFilter.  Disabling range-factorized equivalence classes.
[2020-06-22 22:51:19.066] [jointLog] [info] Usage of --validateMappings implies a default consensus slack of 0.2. Setting consensusSlack to 0.35.
[2020-06-22 22:51:19.066] [jointLog] [info] Using default value of 0.87 for minScoreFraction in Alevin
Using default value of 0.6 for consensusSlack in Alevin
[2020-06-22 22:51:19.250] [alevinLog] [info] Found 88744 transcripts(+455 decoys, +0 short and +0 duplicate names in the index)
[2020-06-22 22:51:19.323] [alevinLog] [info] Filled with 36098 txp to gene entries
[2020-06-22 22:51:19.323] [alevinLog] [error] ERROR: Can't find gene mapping for : PB.40054.21 w/ index 81766
[2020-06-22 22:51:19.323] [alevinLog] [error] ERROR: Txp to Gene Map not found for 52191 transcripts. Exiting
```

What's wrong?

**Comment by roryk** [comment]
Does `PB.40054.21` appear in your `PacBio/single_cell_pipeline/sqanti3_output/Alin_neg_txp2gene.tsv` file? The error is saying it is missing-- is it?

**Comment by rob-p** [comment]
@moschmi,

  @roryk's parsing of the error is exactly what is intended.  Specifically, alevin is looking to map the transcript `PB.40054.21` to a gene, but it is not finding that transcript in the file `PacBio/single_cell_pipeline/sqanti3_output/Alin_neg_txp2gene.tsv`.  The question is a bit trickier when you have your own long-read derived transcriptome, since I don't know the process by which you are grouping transcripts into genes.  It is always possible to have a transcript be its own gene (by having it appear in both the first and second columns in the same line of this file) — however, then you will lose information about how this transcript and other transcripts belonging to the same gene are related, which plays into how alevin quantifies gene-level abundance.

**Comment by moschmi** [comment]
Thanks @roryk and @rob-p 
The transcript PB.40054.21 is in all of the files: The gff files from which I made Alin_neg_txp2gene.tsv, the gentrome.fasta. The gene is "novelGene_36859". All are unique.
 It should be indexed, I have several samples and they all have this error with various transcripts/genes.
What about this error: `ERROR: Txp to Gene Map not found for 52191 transcripts`
The transcript numbers are 88793

**Comment by k3yavi** [comment]
Hi @moschmi ,

Is it possible to share the Alin_neg_txp2gene.tsv file ?

**Comment by rob-p** [comment]
Yes, I wonder if there is some sort of simple formatting issue that could be at fault here.  Having a look at the file(s) would make it much easier to diagnose.

**Comment by moschmi** [comment]
Sure,
attached.
[Alin_neg_txp2gene.zip](https://github.com/COMBINE-lab/salmon/files/4820505/Alin_neg_txp2gene.zip)

**Comment by k3yavi** [comment]
Hi @moschmi ,

Thanks for sharing the file.
You are right the transcript id `PB.40054.21` is indeed present & has the gene mapping in the file you forwarded. Unfortunately, the specific error message was not useful here but the issue is the following: 
![image](https://user-images.githubusercontent.com/8772521/85436584-45706f00-b557-11ea-9ba2-95ebf4e43bc4.png)

If you check the file you forwarded from line number 12,133 - 12,137, it seems a bunch of transcript ids are blank and has no assigned gene-ids. In this case the file parser was not intelligent enough to ignore such empty mappings and end up using the next line (before) tab as the wrong mapped gene-mappings. Later, when we sanity check if the mappings for all transripts were found, it complaints about not being able to find it for a random transcript much lower in the order.

In short, I know you used the bioawk script for making the transcript to gene mapping file, but the script was written with the gencode generated GTF in mind, it seems the one you have has some small difference which is creating the issue. Currently, the easiest fix is to parse the GFF file again and generate the mappings for all the transcript in the proper format.

In future we will add a sanity check for these type of corner cases, thanks for reporting this.

**Comment by moschmi** [comment]
@k3yavi ,
thanks, this could be it. I'll go over my GTF file and let you know

**Comment by moschmi** [comment]
I removed the empty row.
But if I index now the process gets killed when assembling all the "chunks":
`[2020-06-24 14:43:57.001] [puff::index::jointLog] [info] mphf size = 1665.35 MB
[2020-06-24 14:43:57.001] [puff::index::jointLog] [info] chunk size = 313,196,804
[2020-06-24 14:43:57.001] [puff::index::jointLog] [info] chunk 0 = [0, 313,196,804)
[2020-06-24 14:43:57.001] [puff::index::jointLog] [info] chunk 1 = [313,196,804, 626,393,608)
[2020-06-24 14:43:57.001] [puff::index::jointLog] [info] chunk 2 = [626,393,608, 939,590,412)
[2020-06-24 14:43:57.001] [puff::index::jointLog] [info] chunk 3 = [939,590,412, 1,252,787,216)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 4 = [1,252,787,216, 1,565,984,020)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 5 = [1,565,984,020, 1,879,180,824)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 6 = [1,879,180,824, 2,192,377,628)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 7 = [2,192,377,628, 2,505,574,432)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 8 = [2,505,574,432, 2,818,771,236)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 9 = [2,818,771,236, 3,131,968,040)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 10 = [3,131,968,040, 3,445,164,844)
[2020-06-24 14:43:57.002] [puff::index::jointLog] [info] chunk 11 = [3,445,164,844, 3,758,361,611)
Killed`

**Comment by k3yavi** [comment]
I hope this was solved, closing due to inactivity. Please feel free to reopen.
